### PR TITLE
[Feature] [Algorithm] Add FQL Trainer and Configs

### PR DESCRIPTION
**Pull request by neuriv**
## Description

Add `FQLTrainer` for offline pretraining and optional online collection. Register Hydra configs and migrate the recipe to the Trainer. 

Part of #4433. Depends on #4437 and #4438.

<details>
<summary>Training on real OGBench data</summary>

Trained on antmaze-large task 1 using all one million valid offline transitions, four 512-unit hidden layers, batch size 256 and compiled loss. Seed zero; 50 evaluation episodes every 100,000 updates.

image

</details>

<details>
<summary>Training objectives and checkpoint validation</summary>

image

</details>

### CODE REVIEW

File: torchrl/trainers/algorithms/fql.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        if self.collector is None and state_dict["collector"] is not None:
+            self.collector = self.collector_factory()
+        super().load_state_dict(state_dict)
+
+    def shutdown(self) -> None:
+        """Release the optional online collector."""
+        if self.collector is not None:
+            self.collector.shutdown()
+
+    def train(self) -> None:
```
Comment (theap06) [inline torchrl/trainers/algorithms/fql.py:140]: this reimplements the offline+online loop that OfflineToOnlineTrainer and its RB hooks already cover, and it cant take a logger or checkpoint like every other trainer. can FQLTrainer build on those instead of overriding train()?

File: torchrl/trainers/algorithms/fql.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+            progress_bar=False,
+            auto_log_optim_steps=False,
+        )
+        self.offline_steps = offline_steps
+        self.device = device
+        self.loss_call = (
+            torch.compile(loss_module, fullgraph=True) if compile_loss else loss_module
+        )
+        self.replay_keys = [
+            key for key in replay_buffer.storage[0].keys(True, True) if key != "index"
```
Comment (theap06) [inline torchrl/trainers/algorithms/fql.py:102]: storage[0] raises if the buffer is empty at construction; can this be read lazily on first extend?

File: torchrl/trainers/algorithms/fql.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+
+class FQLOptimizationStepper(OptimizationStepper):
+    """Update target critics from the pre-optimizer parameters, as in reference FQL."""
+
+    def step(self, trainer: Trainer, sub_batch: TensorDictBase) -> TensorDictBase:
+        losses = trainer.compute_loss(sub_batch)
+        trainer.optimizer.zero_grad(set_to_none=True)
+        sum(losses[key] for key in trainer.loss_module.out_keys).backward()
+        trainer.target_net_updater.step()
```
Comment (theap06) [inline torchrl/trainers/algorithms/fql.py:27]: is the pre-optimizer target update measurable vs the standard SoftUpdate-after-step order? if not id drop the custom stepper, it only exists to copy a quirk of the jax ref.

File: test/test_fql_trainer.py [hunk 4]
```diff
@@ -0,0 +1,379 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (theap06) [inline test/test_fql_trainer.py:1]: trainer tests should go in test/test_trainer.py rather than a new file.

**Review by theap06 — CHANGES_REQUESTED** [review]

File: test/test_trainer.py [hunk 5]
```diff
@@ -3573,3 +3589,430 @@ def test_rejects_loss_outputs_without_an_optimization_objective(self):
...
+    finally:
+        torch.set_float32_matmul_precision(previous)
+
+
+@pytest.mark.parametrize(
+    "device",
+    [
+        "cpu",
+        pytest.param(
+            "cuda",
```
Comment (theap06) [inline test/test_trainer.py]: we still need this tag pytest.mark.gpu for the cuda ci to be triggered

**Review by theap06 — CHANGES_REQUESTED** [review]
left some feedback. thanks
